### PR TITLE
Add loop scanner to tool-scanner

### PR DESCRIPTION
**Pull request by qinheping**
This extend #3120 with loop scanner that counts the number of loops in each function and the number of functions that contain loops.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: tools/scanner/src/analysis.rs [hunk 1]
```diff
@@ -461,10 +494,23 @@ impl FnLoops {
...
+    }
+
     fn visit_terminator(&mut self, term: &Terminator, location: Location) {
+        // A goto is identified as a loop latch if its target number is less than
+        // the number of visited basic block.
+        if let TerminatorKind::Goto { target } = &term.kind {
+            if self.num_visited_blocks > *target {
+                self.props.loops += 1;
+            }
+        }
```
Comment (celinval) [inline tools/scanner/src/analysis.rs:513]: This is not true for MIR though, you need to traverse the graph.

Comment (qinheping) [reply tools/scanner/src/analysis.rs]: I changed it to find all goto with target that has been visited.

Comment (qinheping) [reply tools/scanner/src/analysis.rs]: I also add the visitor for other terminators that also contains targets beyond GOTO.

File: tools/scanner/src/analysis.rs [hunk 2]
```diff
@@ -461,50 +494,94 @@ impl FnLoops {
...
-                            "rposition",
-                            "nth",
-                            "count",
-                            "last",
-                            "find",
-                        ]
-                        .contains(last)
-                    {
-                        self.props.iterators += 1;
+        match &term.kind {
```
Comment (celinval) [inline tools/scanner/src/analysis.rs:507]: I think you can simplify this by calling `successors()` in the terminator before the match statement: https://github.com/rust-lang/rust/blob/4fe1e2bd5bf5a6f1cb245f161a5e9d315766f103/compiler/stable_mir/src/mir/body.rs#L130

**Review by celinval — APPROVED** [review]
Thanks

**Pull request merged**